### ISSUE TITLE
Add ability to fail fast on certain exception types

### ISSUE THREAD

**Issue by mookid8000** [rebus-org/Rebus#657]
Rebus 1 had the ability to have the number of delivery attempts configured for individual exception types, such that e.g. `ConcurrencyException` would retry 10 times and `ArgumentException` would not retry at all.

This was o.k. in some ways, but it raised a bunch of annoying questions like e.g. "what to do with different exceptions", etc.

Now that I have been asked "can Rebus retry a varying number of times, depending on the exception type?" 100 times, and each and every time it turned out to be because it doesn't make sense to retry at all for some specific exception types – so in fact it wasn't about configuring the number of times, it was more about the ability to fail fast on specific exception types.

Therefore: Add a pipeline step that disables retries for specific exception types, which can be enabled e.g. like so:
```csharp
.Options(o => {
    o.DisableRetriesFor<DomainException>();
});
```
or something like that.

(am unsure about the syntax for disabling the retries.... it's a little bit ugly that it's directly on the `OptionsConfigurer`, but I don't know where else to put it....)

**Comment by boyanio** [comment]
WCF has `FaultException<>` to wrap some faults from the service, so one could introduce a general exception wrapper, f.x. `FailFastException<TException>`, that automatically fails fast.

Furthermore, to allow custom logic for some other exceptions that should fail fast, one could introduce a simple interface, f.x. `IFailFastExceptionHandler` with one method `bool ShouldFailFast(Exception e)` to determine if an exception should lead to fail-fast or retry multiple times. This approach, I believe, is more flexible as it allows you to check not only on the exception type, but also on the exception data.

**Comment by mookid8000** [comment]
@boyanio I like the `FailFastException<>` approach! And I also like what you said about the `IFailFastExceptionHandler`.

I am thinking that the cleanest way of making this work, would be to roll back everything (like when an ordinary exception is caught), and then tell the `IErrorTracker` that the message has reached the limit for how many times it is to be attempted to be consumed.

This would make the `SimpleRetryStrategyStep` forward the message to the error queue when it is received shortly afterwards.

This reason it requires these two steps, is if the user has a unit of work with e.g. database transactions etc., then it is important that these things are rolled back properly – therefore, it is not possible to simply forward the message in the same transaction context.

**Comment by boyanio** [comment]
@mookid8000 I see that `SimpleRetryStrategyStep` gets registered via `IRetryStrategy`, which means theoretically that I, as a user, can define my own retry strategy. So there are two ways to do it conceptually:

- Fail-fast logic happens independently of the retry step (whatever it is), i.e. it aborts the transaction and moves the message to the error queue directly without giving possibility to the retry step to do anything with the message anymore
- Fail-fast logic happens conceptually together with `SimpleRetryStrategyStep` - either extending the class directly or as some kind of decoration

What do you think?

**Comment by mookid8000** [comment]
Either way, as far as I can tell, it would require the ability to immediately forward the failed message to the error queue, but unfortunately that is not possible.

The reason is that the receive operation can be part of a queue transaction where other things have happened, which must be rolled back.

An example could be with MSMQ, where this is the order of operation:

1. `MessageQueueTransaction` is started
1. Message is received
1. Other outbound messages are sent
1. Some kind of exception happens 😞 

In this case, the `MessageQueueTransaction` should be rolled back, in order to "take back" all outbound messages that were sent – but that also rolls back the receive operation, making the message visible to be consumed again.

If the `MessageQueueTransaction` was NOT rolled back, we would risk some nasty side effects caused by outbound messages "escaping" a failed unit of work.

Therefore, I propose the following solution:

* Add decorator for [`IErrorTracker`](https://github.com/rebus-org/Rebus/blob/master/Rebus/Retry/IErrorTracker.cs) with ability to intercept calls to `HasFailedTooManyTimes` and return `true` for a specific message ID
* Add incoming pipeline step right after `SimpleRetryStrategyStep` that catches exceptions and invokes the `ShouldFailFast` callback with a signature something along the lines of `Func<Exception, bool>` which then maybe tells the aforementioned decorator to fail fast for that particular message ID

This will of course result in rolling back everything completely, and then ending up receiving the message again – and only THEN are we able to forward it to the error queue (which will happen automatically, because the error tracker will now report that the message has failed too many times).

Is it somewhat clear now what I mean? 😁

**Comment by mookid8000** [comment]
Fixed by #664 👍
